Add deactivate and uninstall cleanup process

ns-rx opened the pull request:
1. Updated poetwp.php:

- Added a deactivation hook using register_deactivation_hook().
- Implemented the poetwp_deactivate() function, which:
  - Removes the custom 'poem' post type using unregister_post_type().
  - Flushes rewrite rules using flush_rewrite_rules().

2. Added uninstall.php:

* This file handles the plugin uninstallation process.
* It deletes all poems (custom post type 'poem') from the database when the plugin is uninstalled.
* It uses get_posts() to retrieve all poems and wp_delete_post() to remove them.

Resolves #20

ns-rx committed:
Add deactivate and uninstall cleanup process #20

GaryJones commented on uninstall.php:
```diff
@@ -0,0 +1,28 @@
+}
+
+/**
+ * Deletes all poems from the database when the plugin is uninstalled.
```
I'd highly advise NOT deleting user's content as users will not be happy! 

Only delete the demo poem that the plugin added if it exists.

GaryJones replied:
You could have an admin screen with an option to "Delete ALL poems on plugin uninstall." If this option is enabled at the point of uninstalling, go ahead with the full removal—that way, you've got positive buy-in from the user.

The use case here is someone trying multiple plugins to manage their poems rather than wanting the underlying content to be deleted as they try the plugins!

ns-rx replied:
I like this idea, I will add a new issue

GaryJones commented on uninstall.php:
```diff
@@ -0,0 +1,28 @@
+ */
+$args = array(
+	'post_type'      => 'poem',
+	'posts_per_page' => -1,
```
Aside: This will not scale on a large site with many poems if you were to try and delete all of them.

ns-rx committed:
Fix conditional checkfor demo poem

ns-rx committed:
Only delete demo poem, not all poems

ns-rx commented:
1. Fix demo poem title https://github.com/ns-rx/poetwp/pull/37/commits/d0bafe22094714fa851a4c52ea0564826e0cc4b3
   - Updated the demo poem title in demo-poem.php:21
   - Ensures the conditional check for existing demo poem works correctly
   - Accidentally used the wrong branch
Fixes #38 

2. Refine uninstall process https://github.com/ns-rx/poetwp/pull/37/commits/a47b0619013e62544d18f8f973f7fe0d70907f28
   - Modified uninstall.php:15-29 to only delete the demo poem on uninstall
   - Preserves user-generated content while cleaning up plugin-created demo content
Resolves #20

GaryJones commented on includes/demo-poem.php:
```diff
@@ -18,7 +18,7 @@ function poetwp_create_demo_poem() {
 			'post_type'      => 'poem',
 			'post_status'    => 'any',
-			'title'          => 'Demo Poem',
+			'title'          => 'Demo Poem: Windswept',
```
Why not create a variable (for the title, but also maybe the type and status) so that the values used in the WP_Query are the ones used to save the poem. Use the variable in both places.

Remember, keep the data separate from what you do with it.

GaryJones commented on uninstall.php:
```diff
@@ -13,16 +13,17 @@
 	'post_type'      => 'poem',
-	'posts_per_page' => -1,
 	'post_status'    => 'any',
+	'title'          => 'Demo Poem: Windswept',
```
This is now the second (or third) place where this important title is found.

Consider making a value object class that holds this and can be referenced where needed. Something like:

```php
class Demo_Poem {
	public static $post_title = 'Demo Poem: Windswept';
	public static $post_status = 'draft';
	// post type should be grabbed from elsewhere, since it needs to stay consistent with what's been filtered.
}
```

Then here, refer to it as a statically-accessed property:

```php
	'title' => Demo_Poem::$post_title,
```

The demo poem is a thing, an entity, so lets have all the knowledge we have about it live in one object.

ns-rx replied:
Thanks, this was fun to refactor!

ns-rx committed:
Add demo poem class

GaryJones commented on includes/class-demo-poem.php:
```diff
@@ -0,0 +1,76 @@
+	 *
+	 * @var string
+	 */
+	public static $post_type = 'poem';
```
Remember to update (remove) this one when you have a CPT class method that can be referenced instead.

GaryJones commented on includes/class-demo-poem.php:
```diff
@@ -0,0 +1,76 @@
+	 *
+	 * @var string
+	 */
+	public static $post_title = 'Demo Poem: Windswept';
+
+	/**
+	 * The status of the demo poem.
+	 *
+	 * @var string
+	 */
+	public static $post_status = 'draft';
```
You could consider making these fields private and have getter methods for them so that the values can't be set/overridden.

Alternatively, consider making them [readonly properties](https://php.watch/versions/8.1/readonly).

GaryJones commented on includes/class-demo-poem.php:
```diff
@@ -0,0 +1,76 @@
+ *
+ * The `Demo_Poem` class provides static methods to retrieve the content and custom fields for a demo poem.
+ */
+class Demo_Poem {
```
Since we wouldn't want anyone to extend this class, consider making it final:

```suggestion
final class Demo_Poem {
```

GaryJones commented on includes/class-demo-poem.php:
```diff
@@ -0,0 +1,76 @@
+	 *
+	 * @return string The content of the demo poem.
+	 */
+	public static function get_content() {
```
```suggestion
	public static function get_content(): string {
```

GaryJones commented on includes/class-demo-poem.php:
```diff
@@ -0,0 +1,76 @@
+	 *
+	 * @return array An array of custom fields for the demo poem.
+	 */
+	public static function get_custom_fields() {
```
```suggestion
	public static function get_custom_fields(): array {
```

GaryJones commented on includes/class-demo-poem.php:
```diff
@@ -0,0 +1,76 @@
+	/**
+	 * Get custom fields for the demo poem.
+	 *
+	 * @return array An array of custom fields for the demo poem.
```
```suggestion
	 * @return array<string, string> An array of custom fields for the demo poem.
```

We can provide a little more info to IDEs and QA tools by defining what we know the [_array shape_](https://phpstan.org/writing-php-code/phpdoc-types) to be that is returned.

GaryJones commented on includes/demo-poem.php:
```diff
@@ -7,6 +7,11 @@
  * @package    Poetwp\Includes
  */
 
+/**
+ * Includes the Demo Poem class file.
+ */
+require_once POETWP_DIR . 'includes/class-demo-poem.php';
```
Keep all file loading in the root plugin file - that's where this code is bootstrapped, and it's simpler to find it all in one place (in the future, an autoloader can be swapped in).

GaryJones commented on includes/class-demo-poem.php:
```diff
@@ -0,0 +1,76 @@
+ * @since      1.0.0
+ * @package    Poetwp\Includes
+ */
+
```
This file isn't namespaced, which means the `Demo_Poem` class is sitting in the global space.

Consider adding a namespace to this file, and then update the files that reference it with a `use` import statement.

GaryJones commented on uninstall.php:
```diff
@@ -12,13 +12,23 @@
+ */
+require_once 'poetwp.php';
+
+/**
+ * Includes the Demo Poem class file.
+ */
+require_once POETWP_DIR . 'includes/class-demo-poem.php';
```
This wouldn't be needed if the require was already in the main plugin file.

GaryJones commented:
Good job on the refactoring! 

Now that you've got a Demo_Poem object, let's take a step back and think what we're doing with it. We're:

- Adding it to the DB, by checking it exists, and if not, inserting it.
- Removing it.

These are two separate behaviours (services), so you could have a class that handles each aspect (`Demo_Poem_Inserter`, `Demo_Poem_Remover`), or, a class that manages all aspects of it (`Demo_Poem_Manager`).

Although `includes/demo-poem.php` is much smaller now, imagine being able to write code as:

```php
$demo_poem_manager = new Demo_Poem_Manager( new Demo_Poem() );
$demo_poem_manager->maybe_insert();
```

That could be in the root plugin file. The `maybe_insert()` method calls a `poem_exists()` method to check if the demo poem exists, and then it calls an `insert()` method, which does the actual inserting.

Likewise (though I can't remember how much of a plugin is still loaded in memory at the time of an uninstall), the remover could be:

```php
$demo_poem_manager = new Demo_Poem_Manager( new Demo_Poem() );
$demo_poem_manager->remove();
```

Two lines, intuitively named to describe what's happening, without having to understand the args for `wp_delete_post()`, etc.

GaryJones commented:
In terms of removing it - the uninstall / future remove() method is doing a wp_delete_post() call with args that involve the **post_type**, **post_status**, and **title**.

The **post_type** uses the Demo_Poem public property (and will use the CPT class method that can have a filter applied). But what if it was installed with `poem`, and then filtered to be `gary_poem`? Your removal look up will fail to find the demo poem.

The **post_status** is `any`, which is good - it means you're already considering the possibility that someone might have changed the status about the demo poem, which is what this whole comment is about.

The **title** is also a string that the user could change, so a look up of `Demo_Poem::$post_title()` or `Demo_Poem::get_title()` may fail if the title is now something else.

How do we solve this?

{I have an answer, but I'll let you think and suggest first}
